Camera2Basic/utils/src/main/java/com/example/android/camera/utils/YuvToRgbConverter.kt

```diff
@@ -81,118 +79,174 @@ class YuvToRgbConverter(context: Context) {
-            // Y Plane            U Plane    V Plane
-            // ===============    =======    =======
-            // Y Y Y Y Y Y Y Y    U U U U    V V V V
-            // Y Y Y Y Y Y Y Y    U U U U    V V V V
-            // Y Y Y Y Y Y Y Y    U U U U    V V V V
-            // Y Y Y Y Y Y Y Y    U U U U    V V V V
-            // Y Y Y Y Y Y Y Y
-            // Y Y Y Y Y Y Y Y
-            // Y Y Y Y Y Y Y Y
-            val outputStride: Int
-
-            // The index in the output buffer the next value will be written at
-            // For Y it's zero, for U and V we start at the end of Y and interleave them i.e.
-            //
-            // First chunk        Second chunk
-            // ===============    ===============
-            // Y Y Y Y Y Y Y Y    V U V U V U V U
-            // Y Y Y Y Y Y Y Y    V U V U V U V U
-            // Y Y Y Y Y Y Y Y    V U V U V U V U
-            // Y Y Y Y Y Y Y Y    V U V U V U V U
-            // Y Y Y Y Y Y Y Y
-            // Y Y Y Y Y Y Y Y
-            // Y Y Y Y Y Y Y Y
-            var outputOffset: Int
-
-            when (planeIndex) {
-                0 -> {
-                    outputStride = 1
-                    outputOffset = 0
-                }
-                1 -> {
-                    outputStride = 2
-                    // For NV21 format, U is in odd-numbered indices
-                    outputOffset = pixelCount + 1
-                }
-                2 -> {
-                    outputStride = 2
-                    // For NV21 format, V is in even-numbered indices
-                    outputOffset = pixelCount
-                }
-                else -> {
-                    // Image contains more than 3 planes, something strange is going on
-                    return@forEachIndexed
+    private fun Allocation.copyFromPlane(plane: Image.Plane) {
+        if (plane.buffer.hasArray())
+            this.copyFrom(plane.buffer.array())
+        else {
+            this.copyFromBuffer(plane.buffer)
+        }
+    }
+
+    private fun Allocation.copyFromBuffer(buffer: ByteBuffer) {
+        buffer.rewind()
+        val array = ByteArray(buffer.limit())
+        buffer.get(array)
+        this.copyFrom(array)
+    }
+
+    // see https://stackoverflow.com/a/52740776/192373
+// for 1920x1080 interleaved time is reduced from 13 ms to 2 ms
+// for less optimal resolution 1440x1080 to 5 ms
+    private fun Image.toByteArray(outputBuffer: ByteArray) {
+
+        assert(format == ImageFormat.YUV_420_888)
+        assert(planes[1].pixelStride == planes[2].pixelStride)
+        assert(planes[1].rowStride == planes[2].rowStride)
+
+        planes[0].extractLuminance(cropRect, outputBuffer)
+
+        if (planes[2].pixelStride == 2 && planes[2].buffer.isInterleavedWith(planes[1].buffer))
+            planes[2].extractChromaInterleaved(planes[1], cropRect, outputBuffer)
+        else if (planes[2].pixelStride == 2 && planes[1].buffer.isInterleavedWith(planes[2].buffer))
```

Comment:
Can you help me figure it out: if pixelStride == 2, it already means uv interleaved, so don't you need this extra condition? Actually, when do you need extra byte from u-buffer? For me `extractChromaInterleaved` works perfectly fine with only v-buffer as input. Do I miss something?

Reply:
1. if `pixelStride == 2` the two planes *may be* interleaved. Still, I must prove it (that's done by flipping one byte). There is no guarantee .
2. Even when we know that the planes are interleaved, we must find whether it is `uv` or `vu` kind of interleave.
3. On some devices, the length of the **v** ByteBuffer may be `w*h/2`, but on other devices, it is `w*h/2-1`. The missing pixel must be extracted from the second plane.

Reply:
I'm not quite sure, but this code should be tested well, and seems that author is confident about pixelStride == 2.
https://github.com/opencv/opencv/blob/master/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java
Thank you for explanation. 
P.S. tested your code on some devices from our farm with all androids available: 5.1, 6, 7, 8, 9, 10, 11, everything works fine.

Reply:
OpenCV uses C++ to access U and V planes. Here I only have Java at my disposal.